Add heartbeat to etcd quorum check

tgraf opened the pull request:
Depends on #12427 

Adds a heartbeat written to a key in an interval (1min) by the operator. Each etcd client installs a watcher to watch the heartbeat key. When the heartbeat is not updated in 2*interval, the quorum check will start failing:

```
KVStore:                Ok   etcd: 1/1 connected, lease-ID=29c6732d5d580cb5, lock lease-ID=29c6732d5d580cb7, has-quorum=2m2.778966915s since last heartbeat update has been received, consecutive-errors=1: https://192.168.33.11:2379 - 3.4.9 (Leader)
```

When enough consecutive errors have accumulated, the kvstore subsystem will start failing:
```
KVStore:                Failure   Err: quorum check failed 8 times in a row: 4m28.446600949s since last heartbeat update has been received
```

qmonnet commented on pkg/kvstore/etcd.go:
```diff
@@ -974,10 +983,14 @@ func (e *etcdClient) statusChecker() {
-			e.latestErrorStatus = fmt.Errorf("not able to connect to any etcd endpoints")
+			err := fmt.Errorf("not able to connect to any etcd endpoints")
+			e.latestErrorStatus = err
+			e.statusCheckErrors <- err
```
[Travis](https://travis-ci.com/github/cilium/cilium/jobs/359629928) panicked here because it tried to send on a closed channel. On a quick look I couldn't find why the channel is closed, though :/.

```
...
panic: send on closed channel
goroutine 261 [running]:
github.com/cilium/cilium/pkg/kvstore.(*etcdClient).statusChecker(0xc0004fc9c0)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/kvstore/etcd.go:993 +0x7ae
created by github.com/cilium/cilium/pkg/kvstore.connectEtcdClient
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/kvstore/etcd.go:656 +0x62b
FAIL	github.com/cilium/cilium/pkg/kvstore	1.980s
FAIL
Makefile:210: recipe for target 'unit-tests' failed
make: *** [unit-tests] Error 1
The command "./.travis/build.sh" exited with 2.
```

tgraf replied:
I think it's a race condition when the etcd session closes, the channel can already be closed before the status checker is stopped.

tgraf replied:
Resolved by closing the channel when the status checker exits.

qmonnet commented on Documentation/troubleshooting.rst:
```diff
@@ -546,6 +546,82 @@ Opening the ``status``, we can drill down through ``policy.realized.l4``. Do
+  Number of total etcd endpoints and how many of them are reachable.
+
+lease-ID:
+  UUID of the lease used for all keys owned by this agent
```
```suggestion
  UUID of the lease used for all keys owned by this agent.
```

qmonnet commented on Documentation/troubleshooting.rst:
```diff
@@ -546,6 +546,82 @@ Opening the ``status``, we can drill down through ``policy.realized.l4``. Do
+  UUID of the lease used for all keys owned by this agent
+
+lock lease-ID:
+  UUID of the lease used for locks acquired by this agent
```
```suggestion
  UUID of the lease used for locks acquired by this agent.
```

qmonnet commented on Documentation/troubleshooting.rst:
```diff
@@ -546,6 +546,82 @@ Opening the ``status``, we can drill down through ``policy.realized.l4``. Do
+consecutive-errors:
+  Number of consecutive quorum errors. Only printed if errors are present.
+
+https://192.168.33.11:2379 - 3.4.  9 (Leader):
```
```suggestion
https://192.168.33.11:2379 - 3.4.9 (Leader):
```

sayboras commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -251,6 +251,32 @@ func (rc *remoteCluster) onInsert(allocator RemoteIdentityWatcher) {
+	wait:
+		time.Sleep(100 * time.Millisecond)
+		for {
+			rc.mutex.Lock()
```
more about a question, below is the godoc for mutex, I am not sure if we need lock here for accessing _backend_.

If it's required, will ReadLock suffices ?

```
	// mutex protects the following variables
	// - store
	// - remoteNodes
	// - ipCacheWatcher
	// - remoteIdentityCache
	mutex lock.RWMutex
```

tgraf replied:
The documentation is wrong, fixed. Changed to readlock.

sayboras commented on pkg/kvstore/backend.go:
```diff
@@ -138,6 +141,10 @@ type BackendOperations interface {
 
+	// StatusCheckErrors returns a channel which receives status check
+	// errors
+	StatusCheckErrors() chan error
```
nits: is this channel _read only_ ?

sayboras commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -251,6 +251,32 @@ func (rc *remoteCluster) onInsert(allocator RemoteIdentityWatcher) {
+			}
+
+			select {
+			case err := <-statusCheckErrors:
```
nit: select can be removed as there is only one _case_

tgraf replied:
This wasn't meant to also check `rc.changed` but it got lost. Fixed.

joestringer commented on operator/kvstore_watchdog.go:
```diff
@@ -88,4 +88,14 @@ func startKvstoreWatchdog() {
+
+	go func() {
+		for {
+			err := kvstore.Client().Update(context.Background(), kvstore.HeartbeatPath, []byte(time.Now().Format(time.RFC3339)), true)
```
Is `context.Background()` OK here or can it get stuck if for instance we go from connected -> unconnected for a period -> connected again?

Basically wondering whether there should be a timed context here instead with `HeartbeatWriteInterval` to retry.

joestringer commented on operator/kvstore_watchdog.go:
```diff
@@ -88,4 +88,14 @@ func startKvstoreWatchdog() {
 		}
 	}()
+
+	go func() {
+		for {
```
Should this be a controller instead so it's easy to see the status of this logic from `cilium status` and not just logs?

tgraf replied:
It's running as part of the operator which does not have the controller framework for visibility. We can change that at some point. The operator is full of go routines right now.

joestringer commented on pkg/kvstore/kvstore.go:
```diff
@@ -41,4 +45,12 @@ const (
+
+	// HeartbeatPath is the path to the key at which the operator updates
+	// the heartbeat
+	HeartbeatPath = BaseKeyPrefix + "/.heartbeat"
```
It would be good to include this also in the new leases docs https://docs.cilium.io/en/latest/kvstore/#leases

joestringer commented on pkg/kvstore/etcd.go:
```diff
@@ -608,6 +613,31 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
+			select {
+			case _, ok := <-watcher.Events:
+				if !ok {
+					watcher.Stop()
+					return
```
Does this need logging or do we consider this to be the "normal exit" path?

Just wondering whether we could end up in a situation where this goroutine exits prematurely and how we might introspect such a case.

tgraf replied:
It should never exit prematurely but added a log just to be safe.

joestringer commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -251,6 +251,32 @@ func (rc *remoteCluster) onInsert(allocator RemoteIdentityWatcher) {
 		}
 	}()
+
+	go func() {
```
Could/should this be a controller too?

tgraf replied:
Yes it can, changed.

joestringer commented on Documentation/troubleshooting.rst:
```diff
@@ -546,6 +546,82 @@ Opening the ``status``, we can drill down through ``policy.realized.l4``. Do
+etcd health and potentially take action. The interval depends on the overall
+cluster size. The larger the cluster, the longer the interval:
+
+ * If no etcd endpoints can be reached, Cilium will report failure in ``cilium
+   status``. This will cause the liveness and readiness probe of Kubernetes to
+   fail and Cilium will be restarted.
```
I was wondering whether this basically means that the datapath is tied to the control plane more closely, in that an etcd outage could cause connectivity outages. But then I remembered that the cilium agent is not really the datapath, it's part of the control plane so even if there is an extended etcd outage and the cilium agent is out-of-action for some time in a `Crashloopbackoff` state, connectivity should remain.

Whether we test/validate this is perhaps another question (Shall we file an issue for this?).

Is it worth mentioning this point anywhere here, do you think? That dataplane connectivity for existing endpoints should continue to operate during such an outage?

tgraf replied:
We currently have [0] which is fairly outdated as well. We can of course improve it.

We are not introducing significant downtime of the agent but it will result in proxies being restarted. This was already possible before this PR when no etcd endpoint was reachable so a total etcd outage always resulted in agent restarts. We are extending the definition of etcd outage to also including loss of quorum over a long period of time and failure to receive updates for a long period of time.

[0] https://docs.cilium.io/en/v1.8/concepts/failure_behavior/

joestringer replied:
Looks like this is resolved by folding the linked doc into this troubleshooting section.

joestringer commented on Documentation/troubleshooting.rst:
```diff
@@ -546,6 +546,82 @@ Opening the ``status``, we can drill down through ``policy.realized.l4``. Do
+
+In addition, Cilium performs a background check in an interval to determine
+etcd health and potentially take action. The interval depends on the overall
+cluster size. The larger the cluster, the longer the interval:
```
Do you think it would be useful to link to this to help developers supporting this to understand where to start on inspecting this area?

https://pkg.go.dev/github.com/cilium/cilium/pkg/kvstore?tab=doc#ExtraOptions.StatusCheckInterval

(The formatting is a little messed up in pkg.go.dev but we can fix that up)

joestringer commented on Documentation/troubleshooting.rst:
```diff
@@ -546,6 +546,82 @@ Opening the ``status``, we can drill down through ``policy.realized.l4``. Do
+  particular endpoint is currently the elected leader. If an etcd endpoint
+  cannot be reached, the error is shown.
+
+Recovery behavior
+-----------------
```
What's the overlap between this and https://docs.cilium.io/en/v1.8/concepts/failure_behavior/ ? Should we cross-link them?

tgraf replied:
I've merged the two together.

joestringer commented on Documentation/troubleshooting.rst:
```diff
@@ -546,6 +546,82 @@ Opening the ``status``, we can drill down through ``policy.realized.l4``. Do
+   quorum. If this operation fails, loss of quorum is reported. If quorum fails
+   for three or more intervals in a row, Cilium is declared unhealthy.
+
+ * The Cilium operator will constantly write to a heartbeat key
+   (``cilium/.heartbeat``). All Cilium agents will watch for updates to this
+   heartbeat key. This validates the ability for an agent to receive key
+   updates from etcd. If the heartbeat key is not updated in time, the quorum
+   check is declared to have failed and Cilium is declared unhealthy after 3 or
+   more consecutive failures.
```
Do we still have a location in the docs that describes the various duties that cilium-operator has? Would be good to list its impact on Cilium startup / restart due to this aspect if so.

tgraf replied:
What are you looking for in particular? Restarts don't matter. Only if the operator is unavailable for a long period of time then complications will start to occur. That's already the case today.  

We currently have:
https://docs.cilium.io/en/v1.8/concepts/overview/#cilium

joestringer replied:
I was thinking that the tolerance for operator being out-of-action is much higher today, depending on cluster size/activity it could be hours then starts failing in weird ways. AFAIK with this new logic we're looking at more like several minutes before Cilium starts restarting, though I think it will be more straightforward to understand/debug with this new behaviour.

(I also realize now that we'll need to consider how this impacts #12409 )

Probably we can just tweak the end of the sentence in the docs link you provided to make it more clear how operator unavailability may impact the cluster with this new logic.

tgraf replied:
I've tweaked it further. We should consider that this PR is targeted for backports so at some point we'll have to draw a line at how many small details we touch.

joestringer reviewed (changes requested):
Metrics side looks good. I have a few questions around introspection of the new state and a few around docs enhancements.

AFAIK in the event of a complete etcd cluster loss, Cilium still retains a 15-minute timeout on startup where it attempts to establish a healthy connection to an etcd cluster with quorum before exiting. This aspect is not really covered in this PR given that it's more focused on partial outages, but just wondering whether you had been looking at "complete etcd outage" type scenarios with the new logic in mind? At a glance I think it should be orthogonal, no specific concerns come to mind about how that case might be impacted by this new logic.

tgraf committed:
doc: Document etcd failure behavior

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
clustermesh: Disable initlock quorum check

Clustermesh is never performing write operations so the lock-based
quorum check is only adding contention to remote etcds.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kvstore: Add hearbeat to detect stale etcd connections

Adds a heartbeat written to a key in an interval (1min) by the operator.
Each etcd client installs a watcher to watch the heartbeat key. When the
heartbeat is not updated in 2*interval, the quorum check will start
failing:

```
KVStore:                Ok   etcd: 1/1 connected, lease-ID=29c6732d5d580cb5, lock lease-ID=29c6732d5d580cb7, has-quorum=2m2.778966915s since last heartbeat update has been received, consecutive-errors=1: https://192.168.33.11:2379 - 3.4.9 (Leader)
```

When enough consecutive errors have accumulated, the kvstore subsystem
will start failing:
```
KVStore:                Failure   Err: quorum check failed 8 times in a row: 4m28.446600949s since last heartbeat update has been received
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
clustermesh: Improve error message for inital connection attempt

"Backend not initialized" does not mean much to users.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
clustermesh: Restart etcd connection on quorum errors

Watch the status of the etcd conection and restart the connection if
quorum loss is detected. Given that lock acquisition is disabled for
clustermesh, the quorum check equals to the ability to receive updates
on the heartbat key.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
clustermesh: Fix comments on scope of rc.mutex

Reported-by: @sayboras
Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kvstore: Add metric to count quorum errors

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
kvstore: Improve initial status message

The initial status message of the etcd subsystem is:

```
KVStore:                Ok   No connection to etcd
```

This can be misleading as it does not indicate whether the etcd session
was ever established or not. Clarify this:

```
KVStore:                Ok   Waiting for initial connection to be established
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on Documentation/kvstore.rst:
```diff
@@ -103,6 +103,20 @@ Key                                                              Value
+====================== ======================
+Key                    Value
+====================== ======================
+``cilium/.heartbeat``  Current time and date
```
This is governed by `KVstoreLeaseTTL` as well, right? We could also add a line to the lease expiry table about 20 lines down.

joestringer commented on Documentation/troubleshooting.rst:
```diff
@@ -546,6 +546,130 @@ Opening the ``status``, we can drill down through ``policy.realized.l4``. Do
+  identity allocation which requires interaction with the kvstore. If a
+  workload can be scheduled due to re-using a known security identity, then
+  state propagation of the endpoint details to other nodes will still depend on
+  the kvstore.
```
[0]

joestringer commented on Documentation/troubleshooting.rst:
```diff
@@ -546,6 +546,130 @@ Opening the ``status``, we can drill down through ``policy.realized.l4``. Do
+All datapath operations:
+  All datapath forwarding, policy enforcement and visibility functions do not
+  depend on the kvstore. Packets will continue to be forwarded and network
+  policy rules will continue to be enforced.
```
nit: When running kvstore mode with direct routing, if one is able to schedule a new endpoint due to [0] it's possible for policy drops to occur for peers who would otherwise be able to communicate, because the ipcache translation for gathering the identity would not be possible. I guess this is getting pretty detailed in a corner case though so not sure if there's any reasonable way to cover corner cases like this in the wording.

joestringer replied:
I think the statement is still broadly true and this newly-deployed-endpoints corner case is probably not worth mentioning.

tgraf replied:
I've clarified this further:

>   Scheduling of new workloads:
    As part of of scheduling workloads/endpoints, agents will perform security
    identity allocation which requires interaction with the kvstore. If a
    workload can be scheduled due to re-using a known security identity, then
    state propagation of the endpoint details to other nodes will still depend on
    the kvstore and thus policy drops may be observed as other nodes in the
    cluster will not be aware of the new workload.

joestringer commented on Documentation/troubleshooting.rst:
```diff
@@ -546,6 +546,130 @@ Opening the ``status``, we can drill down through ``policy.realized.l4``. Do
+  bootstrap time, however, the agent will still perform all possible operations
+  while waiting for the kvstore to appear.
+
+Operations which *do not* require kvstore availability:
```
Datapath and NetworkPolicy are mentioned below, presumably even new intra-cluster services would be plumbed correctly?

joestringer commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -251,6 +254,39 @@ func (rc *remoteCluster) onInsert(allocator RemoteIdentityWatcher) {
+
+					err, ok := <-statusCheckErrors
+					if ok && err != nil {
+						rc.getLogger().WithError(err).Warning("Error observed on etcd connection, reconnecting etcd")
```
At the moment this is hiding the errors from the controller infrastructure, which means it's no more visible than before; we'll just see yet another controller running with an "OK" status. Could/should we return an error in this case to get the controller infrastructure to retrigger this logic? Or are we trying to avoid it to keep retries to a short period?

tgraf replied:
I've had to remove the controller again. The controller architecture is not yet useful to listen on channels as it relies on completing runs regularly in order to terminate controllers. I've switched back to a go routine to make backports feasible.

joestringer replied:
I wondered if that might be the conclusion, wasn't sure if we can get away with the controller not triggering for a long time then it will just reschedule once we return an error. But if we need more controller package infrastructure then it makes sense to go back to how it was. Thanks for exploring the idea anyway.

joestringer reviewed (changes requested):
One more main note on the controller change and visibility below, the rest is minor nits that could be addressed here or as a follow up.

tgraf committed:
kvstore: Log errors while closing etcd client

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
clustermesh: Release old connection asynchronously

When releasing the etcd connection, sessions are attempted to be
revoked. In the event of an unhealthy etcd connection, the operation
will fail and time out. This operation will take a long time though.
Instead of blocking, release the resources in the background.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
clustermesh: Report number of failures in status

Good condition:
```
   cluster2: ready, 4 nodes, 3 identities, 1 services, 0 failures (last: never)
```

Bad condition:
```
   cluster2: not-ready, 0 nodes, 0 identities, 0 services, 1 failures (last: 9s ago)
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -1819,6 +1819,13 @@ definitions:
+        description: Number of failures reaching the cluster
+        type: integer
+      last-failure:
+        description: Number of failures reaching the cluster
```
nit:

```suggestion
        description: Time of last failure that occurred while attempting to reach the cluster
```

joestringer commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -107,45 +115,58 @@ func (rc *remoteCluster) getLogger() *logrus.Entry {
+	rc.backend = nil
+	rc.mutex.Unlock()
+
+	// Release resources asynchroneously in the background. Many of these
+	// operations may time out if the connection was closed due to an error
+	// condition.
```
The implementation isn't really changing here but I note that the contexts have no deadline below, and was wondering whether `remoteNodes.Close()` or `remoteServices.Close()` could block indefinitely?

(Could be a follow-up)

joestringer replied:
Note that if the answer is "yes" then we need to deal with the other `context.TODO()` in the controller failure cases below as well...

joestringer replied:
Related: https://github.com/cilium/cilium/pull/12526

tgraf replied:
I had looked at that and while it should have a timeout, it doesn't matter that much for clustermesh as the blocking operation is the deletion of local keys. The SharedStore is only used to listen so no local keys have to be deleted when the store is closed. We can fix it outside of this PR.

joestringer reviewed (changes requested):
One more minor nit in the api, and one question on the `context` usage here.
